### PR TITLE
Restrict argument types having a default value

### PR DESCRIPTION
**Pull request by MakeNowJust**
For example:

```crystal
def foo(x : String = 1)
  x
end

foo
```

I think above code is invalid, but current compiler accepts it.
This pull request fixes it.

Another effect, below code works fine now:

```crystal
def foo(x : T.class = String) forall T
  {{ T }}
end

foo # => String
```

(Before, we got "undefined constant T" error.)

Implementation point is introducing `TypeRestrict` fictitious node and applying it when instantiate a method with arguments having a default value.

### CODE REVIEW

File: src/compiler/crystal/semantic/ast.cr [hunk 1]
```diff
@@ -77,6 +77,23 @@ module Crystal
     def_equals_and_hash type
   end
 
+  # Fictitious node to represent a type restriction
+  #
+  # It is used for type restrection of method arguments.
+  class TypeRestrict < ASTNode
```
Comment (asterite) [inline src/compiler/crystal/semantic/ast.cr]: Maybe call this `TypeRestriction`?

File: src/compiler/crystal/semantic/ast.cr [hunk 2]
```diff
@@ -77,6 +77,23 @@ module Crystal
...
+    getter value
```
Comment (asterite) [inline src/compiler/crystal/semantic/ast.cr]: Maybe call this `node`?

Comment (MakeNowJust) [inline src/compiler/crystal/semantic/ast.cr]: But, if this is `node`, we use `node.node` in `MainVisitor#visit`.

File: src/compiler/crystal/semantic/ast.cr (hunk 1 above)
Comment (MakeNowJust) [inline src/compiler/crystal/semantic/ast.cr]: Ok. I'll fix.

File: src/compiler/crystal/semantic/main_visitor.cr [hunk 3]
```diff
@@ -2836,8 +2834,31 @@ module Crystal
       false
     end
 
+    def visit(node : TypeRestrict)
+      value = node.value
+      restriction = node.restriction
+
+      value.accept self
+
+      unless context = match_context
```
Comment (asterite) [inline src/compiler/crystal/semantic/main_visitor.cr]: I'm thinking, maybe there's a different way to do this. The problem with this approach is that the type is checked once here (below). But what happens if the type of `value` changes later? (rare, but can happen)

Another way to do this is to do something like:

```cr
# This line is currently in `visit(node : Path)`, so we could probably extract it to a helper method
restriction_type = (@path_lookup || @scope || @current_type).lookup_type_var(node, free_vars: @free_vars)
unless restriction_type.is_a?(Type)
  node.raise "restriction must be a type, not #{restriction_type}"
end

# With this we make the node always have this type, and raise an error if it's not (also after a change)
value.freeze_type = restriction_type
```

In this way I think there's no need to change the visitor to have a `MatchContext` instead of `free_vars`.

Could you try this change and see if it works?

**Comment by asterite** [comment]
@MakeNowJust This is excellent, thank you!

I made a comment about a slightly different way of implementing this, but the basic idea you provided is really good. In fact this fixes #2636

File: src/compiler/crystal/semantic/ast.cr [hunk 4]
```diff
@@ -77,6 +77,23 @@ module Crystal
     def_equals_and_hash type
   end
 
+  # Fictitious node to represent a type restriction
+  #
+  # It is used for type restrection of method arguments.
+  class TypeRestriction < ASTNode
+    getter value
```
Comment (asterite) [inline src/compiler/crystal/semantic/ast.cr]: @MakeNowJust Ah, OK. Let's leave it as `value` for now :-)

File: src/compiler/crystal/semantic/ast.cr (hunk 2 above)
Comment (MakeNowJust) [inline src/compiler/crystal/semantic/ast.cr]: `TypeRestriction`'s semantics looks like `Cast`, so I think it is good that these getters rename same as `Cast`.

Comment (asterite) [inline src/compiler/crystal/semantic/ast.cr]: Sounds good 👍

File: src/compiler/crystal/semantic/to_s.cr [hunk 5]
```diff
@@ -44,7 +44,7 @@ module Crystal
       false
     end
 
-    def visit(node : TypeRestrict)
+    def visit(node : TypeRestriction)
       @str << "# type restrict: "
```
Comment (Sija) [inline src/compiler/crystal/semantic/to_s.cr]: leftover typo: `restict` -> `restriction`

Comment (MakeNowJust) [inline src/compiler/crystal/semantic/to_s.cr]: Yes.

File: src/compiler/crystal/semantic/main_visitor.cr (hunk 3 above)
Comment (MakeNowJust) [inline src/compiler/crystal/semantic/main_visitor.cr]: `freeze_type` looks good. But, I think this way doesn't solve `def foo(x : T.class = String)` problem, and fixing this problem needs real call of `restrict` method, so `match_context` is needed.

> The problem with this approach is that the type is checked once here (below). But what happens if the type of value changes later? (rare, but can happen)

Could you show me sample code?

Comment (asterite) [inline src/compiler/crystal/semantic/main_visitor.cr]: If the default value is a method call. But don't worry, this case won't be possible in the future, and right now it's actually very to trigger that. I think your implementation is more than good, so let's leave it like that :-)

**Comment by bcardiff** [comment]
Regarding https://github.com/crystal-lang/crystal/pull/3834#discussion_r94593677, I would allow method calls in default methos so this things still work `def foo(seed = T.default)`.

It make me suspicious that a node (even fictitious one) is added. Could the issue be resolved by enforcing that, when instantiating a def, the default value is_a? value of the type restriction (if any)?

File: src/compiler/crystal/semantic/main_visitor.cr [hunk 6]
```diff
@@ -2836,8 +2834,31 @@ module Crystal
...
+      obj.accept self
+
+      unless context = match_context
+        node.raise "BUG: there is no match context"
+      end
+
+      if type = obj.type.restrict(to, context)
+        node.type = type
+      else
+        node.raise "can't restrict #{obj.type} to #{to}"
```
Comment (asterite) [inline src/compiler/crystal/semantic/main_visitor.cr]: A spec is missing for this (the specs you added seem to be testing other things)

**Comment by Sija** [comment]
yup, 👍  for keeping method calls in default argument values.

```cr
class Foo
  protected def default_name
    preview? && "foo-preview" || "foo"
  end

  def save(name = default_name)
    # ...
  end
end
```

**Comment by MakeNowJust** [comment]
@bcardiff @Sija I think all method calls in default value don't deny, and some pathological method calls deny (I considered such an example, but I can't imagine it.)

**Comment by bcardiff** [comment]
@MakeNowJust Sorry, could you rephrase or expand? I don't follow you.

**Comment by MakeNowJust** [comment]
@bcardiff I don't think all method calls in default value are banned by compiler, so don't worry.

File: src/compiler/crystal/semantic/main_visitor.cr (hunk 6 above)
Comment (MakeNowJust) [inline src/compiler/crystal/semantic/main_visitor.cr]: Yes. I just added.

File: spec/compiler/semantic/def_spec.cr [hunk 7]
```diff
@@ -171,6 +171,16 @@ describe "Semantic: def" do
       "undefined method"
   end
 
+  it "errors when type restriction is incompatible with default value" do
```
Comment (Sija) [inline spec/compiler/semantic/def_spec.cr]: I'd rephrase it: `errors when default value is incompatible with type restriction`

Comment (MakeNowJust) [inline spec/compiler/semantic/def_spec.cr]: ok

**Comment by asterite** [comment]
@MakeNowJust Thank you for this!
